src/middlewared/middlewared/plugins/activedirectory.py

```diff
@@ -0,0 +1,1147 @@
+            if res:
+                self._isopen = True
+            elif saved_gssapi_error:
+                raise CallError(saved_gssapi_error)
+            elif saved_simple_error:
+                raise CallError(saved_simple_error)
+
+        return (self._isopen is True)
+
+    def _close(self):
+        self._isopen = False
+        if self._handle:
+            self._handle.unbind()
+            self._handle = None
+
+    def _search(self, basedn='', scope=ldap.SCOPE_SUBTREE, filter='', timeout=-1, sizelimit=0):
+        if not self._handle:
+            self._open
+
+
+        result = []
+        results = []
+        serverctrls = None
+        clientctrls = None
+        paged = SimplePagedResultsControl(
+            criticality=False,
+            size=self.pagesize,
+            cookie=''
+        )
+        paged_ctrls = { SimplePagedResultsControl.controlType: SimplePagedResultsControl }
+
+        if self.pagesize > 0:
+            page = 0
+            while True:
+                serverctrls = [paged]
+
+                id = self._handle.search_ext(
+                    basedn,
+                    scope,
+                    filterstr=filter,
+                    attrlist=None,
+                    attrsonly=0,
+                    serverctrls=serverctrls,
+                    clientctrls=clientctrls,
+                    timeout=timeout,
+                    sizelimit=sizelimit
+                )
+
+                (rtype, rdata, rmsgid, serverctrls) = self._handle.result3(
+                    id, resp_ctrl_classes=paged_ctrls
+                )
+
+                result.extend(rdata)
+
+                paged.size = 0
+                paged.cookie = cookie = None
+                for sc in serverctrls:
+                    if sc.controlType == SimplePagedResultsControl.controlType:
+                        cookie = sc.cookie
+                        if cookie:
+                            paged.cookie = cookie
+                            paged.size = self.pagesize
+
+                        break
+
+                if not cookie:
+                    break
+
+                page += 1
+        else:
+            id = self._handle.search_ext(
+                basedn,
+                scope,
+                filterstr=filter,
+                attrlist=attributes,
+                attrsonly=attrsonly,
+                serverctrls=serverctrls,
+                clientctrls=clientctrls,
+                timeout=timeout,
+                sizelimit=sizelimit
+            )
+
+            type = ldap.RES_SEARCH_ENTRY
+            while type != ldap.RES_SEARCH_RESULT:
+                try:
+                    type, data = self._handle.result(id, 0)
+
+                except ldap.LDAPError as e:
+                    self._logex(e)
+                    break
+
+                results.append(data)
+
+            for i in range(len(results)):
+                for entry in results[i]:
+                    result.append(entry)
```

Comment:
is results[i] a list? Perhaps it could be `result += results[i]` or `result += list(results[i])`

Reply:
Even better. I realized that this block of code is unnecessary so I just removed it.